From 7c74575030512224a79c3de2dfa4b7ed40115f23 Mon Sep 17 00:00:00 2001
From: Michael Hutchinson <m.j.hutchinson@gmail.com>
Date: Sat, 29 Jun 2013 05:11:25 -0400
Subject: [PATCH] Implement IME reset for imquartz

Reset the Cocoa IME state when the immodule is reset.

This means that IME state is cleared correctly when moving the
caret within a widget, or transferring focus between widgets, and
the preedit string is cleared.
---
 modules/input/imquartz.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/modules/input/imquartz.c b/modules/input/imquartz.c
index 9788ad3..f70399e 100644
--- a/modules/input/imquartz.c
+++ b/modules/input/imquartz.c
@@ -211,6 +211,24 @@ static void
 quartz_reset (GtkIMContext *context)
 {
   GTK_NOTE (MISC, g_print ("quartz_reset\n"));
+  GtkIMContextQuartz *qc = GTK_IM_CONTEXT_QUARTZ (context);
+
+  if (!qc->client_window)
+    return;
+
+  NSView *nsview = gdk_quartz_window_get_nsview (qc->client_window);
+  if (!nsview)
+    return;
+
+  // reset any partial input for this NSView
+  NSInputManager *currentInputManager = [NSInputManager currentInputManager];
+  [currentInputManager markedTextAbandoned:nsview];
+
+  if (qc->preedit_str) {
+    g_free (qc->preedit_str);
+    qc->preedit_str = NULL;
+    g_signal_emit_by_name (context, "preedit_changed");
+  }
 }
 
 static void
@@ -239,6 +257,9 @@ quartz_focus_out (GtkIMContext *context)
 
   GtkIMContextQuartz *qc = GTK_IM_CONTEXT_QUARTZ (context);
   qc->focused = FALSE;
+
+  // clear any partially built strings or it'll mess up other GTK+ widgets in the window
+  quartz_reset (context);
 }
 
 static void
-- 
1.8.2.3

